app/src/main/java/org/mozilla/fenix/perf/StartupTimeline.kt

```diff
@@ -44,4 +51,66 @@ object StartupTimeline {
     private fun advanceState(startingActivity: StartupActivity) {
         state = StartupTimelineStateMachine.getNextState(state, startingActivity)
     }
+
+    /**
+     * Measures the given [measuredFunction] under the given [metric].
+     *
+     * For debug purposes, measured values may be logged to logcat or seen in the Glean Debug View by:
+     * - Enabling telemetry if it's not already enabled
+     * - Following the steps in the Glean docs:
+     *   https://mozilla.github.io/glean/book/user/debugging/android.html
+     *
+     * We prefer to aggregate all measurements through this class, rather than using the metrics
+     * directly, in order to:
+     * - Centralize startup performance analysis in this class
+     * - Be able to quickly find all startup measurements with tooling (i.e. "find usages")
+     * - Easily modify the code that runs for all measurements
+     *
+     * We use Glean for all of our measurements to avoid reimplementing non-trivial measurement code
+     * that already exists and to ensure the values reported by telemetry are the same that are
+     * extracted by our perf test CI.
+     *
+     * Example usage:
+     * ```
+     * import ...Telemetry.geckoRuntimeCreate
+     *
+     * val runtime = StartupTimeline.measure(geckoRuntimeCreate) {
+     *     GeckoRuntime.create()
+     * }
+     * ```
+     *
+     * @return the return value of [measuredFunction].
+     */
+    @Suppress("TooGenericExceptionCaught") // we need to catch everything to cancel correctly.
+    inline fun <R> measure(metric: TimespanMetricType, measuredFunction: MeasuredFunction<R>): R {
+        metric.start()
+
+        val returnValue = try {
+            measuredFunction()
+        } catch (e: Exception) {
+            metric.cancel()
+            throw e
+        }
+
+        metric.stop()
+        return returnValue
+    }
+}
+
+typealias MeasuredFunction<R> = () -> R
+
+/**
+ * A [LifecycleObserver] for [HomeActivity] focused on startup performance measurement.
+ */
+class StartupHomeActivityLifecycleObserver : LifecycleObserver {
+    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
+    fun onStop() {
+        // Startup metrics placed in the Activity should be re-recorded each time the Activity
+        // is started so we need to clear the ping lifetime by submitting once per each startup.
+        // It's less complex to add it here rather than the visual completeness task manager.
+        //
+        // N.B.: this submission location may need to be changed if we add metrics outside of the
```

Comment:
I'm a bit confused by this: metrics are cleared whenever you send a ping (`ping` lifetime). You should only be concerned in sending the ping after the metrics are set: startup only happens once per session anyway, right?

Reply:
This is currently placed in the `HomeActivity.onStop` method but if the user enters the app through a separate Activity and closes the app without seeing HomeActivity, this will never be called. At some point, we'll probably want to measure these alternative use cases.

After we're more invested in startup telemetry measurement, we should probably move it to work for startup of every activity. I didn't want to do the obvious thing right now (call it in onStop of all activities because it's unclear if other activities can start before the previous one stops, which would break the lifecycle).

I added a brief comment to elaborate. Please resolve if this makes sense.